### PR TITLE
ARROW-14114: [C++][Parquet] Fix multi-threaded read of PME files

### PR DESCRIPTION
**Pull request by andersonm-ibm**
Change AesDecryptor to be per Decryptor, instead of shared.
This solves the problem of reading with PME using multiple threads.

Details:
It was discovered when exposing high-level PME in PyArrow that reading an encrypted parquet file in PyArrow intermittently fails decryption finalization and sometime fails with Segmentation fault. The same in C++ reading an encrypted parquet with FileReader.ReadTable() multithreaded (with set_use_threads(true) ).
The current implementation uses two caches: meta_decryptor_ and data_decryptor_ , for AesDecryptors, and every Decryptor gets the same AesDecryptor with AesDecryptorImpl from this cache. 
However, AesDecryptor::AesDecryptorImpl::GcmDecrypt() and AesDecryptor::AesDecryptorImpl::CtrDecrypt() use ctx_ member of type EVP_CIPHER_CTX from OpenSSL, which shouldn't be used from multiple threads concurrently.
So, instead of sharing the same AesDecryptor between all Decryptors, an AesDecryptor will be created per Decryptor, which is per column.

Co-authored-by: Gidon Gershinsky <ggershinsky@apple.com>

CC @thamht4190 @pitrou @revit13

### CODE REVIEW

**Comment by pitrou** [comment]
Can you explain what is being accessed from multiple threads in this context?

**Comment by andersonm-ibm** [comment]
> Can you explain what is being accessed from multiple threads in this context?

`AesDecryptor::AesDecryptorImpl::GcmDecrypt()` and `AesDecryptor::AesDecryptorImpl::CtrDecrypt()` use `ctx_` member of type `EVP_CIPHER_CTX` from OpenSSL, which shouldn't be used from multiple threads concurrently. 
So, instead of sharing the same `AesDecryptor` between all `Decryptors`, an `AesDecryptor` will be created per `Decryptor`, which is per column.

File: cpp/src/parquet/encryption/encryption_internal.h [hunk 1]
```diff
@@ -88,8 +88,9 @@ class AesDecryptor {
   explicit AesDecryptor(ParquetCipher::type alg_id, int key_len, bool metadata,
                         bool contains_length = true);
 
-  static AesDecryptor* Make(ParquetCipher::type alg_id, int key_len, bool metadata,
-                            std::vector<AesDecryptor*>* all_decryptors);
+  static std::shared_ptr<AesDecryptor> Make(
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/parquet/encryption/encryption_internal.h:91]: It seems you forgot to address this comment?

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp -g python

File: cpp/src/parquet/encryption/encryption_internal.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (andersonm-ibm) [reply cpp/src/parquet/encryption/encryption_internal.h:91]: Sorry, I thought if would be enough to add the description of the `all_decryptors `member.

[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/parquet/encryption/encryption_internal.h:91]: (also the `metadata` argument deserves describing)

**Comment by pitrou** [comment]
@andersonm-ibm Can you also rebase on git master?

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp -g python

File: cpp/src/parquet/encryption/encryption_internal.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (andersonm-ibm) [reply cpp/src/parquet/encryption/encryption_internal.h:99]: Sure, @pitrou , I've added the comment. 
Thanks for rebasing.
Please, let me know if I can help with anything else..

**Review by pitrou — APPROVED** [review]
Thanks for the update @andersonm-ibm . This is looking good to me now, will merge.

**Pull request closed without merging**